### PR TITLE
Wireframe view

### PR DESCRIPTION
**Pull request by BagelOrb**
A new view which only shows the edges and shows overhang on the edges.

I haven't gotten edge-overhang to be calculated, because each edge needs to know about both adjoining faces, which is not supported in the RenderBatch, I think.

Instead the edges are colored according to the normal face-overhang :/

The current implementation is only supported by version 4.1...

### CODE REVIEW

**Comment by fieldOfView** [comment]
I don't mean to alarm you, but it looks like you're writing Python

**Comment by thopiekar** [comment]
:fearful:

**Comment by BagelOrb** [comment]
Nah it's just some scripting to get the shader c++ code to work

xD

Op 20 mei 2017 14:53 schreef "Thomas Karl Pietrowski" <
notifications@github.com>:

File: plugins/WireframeView/wireframe.shader [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        gl_Position = u_model_view_projection_matrix * a_vertex;
+    }
+
+fragment =
+    uniform lowp vec4 u_color;
+
+    void main()
+    {
+        gl_FragColor = u_color;
```
Comment (Ghostkeeper) [inline plugins/WireframeView/wireframe.shader]: This gives a silhouette of the object if the user has no OpenGL 4 support.

**Comment by Ghostkeeper** [comment]
Devs, see CURA-3849.

File: plugins/WireframeView/wireframe.shader (hunk 1 above)
Comment (BagelOrb) [inline plugins/WireframeView/wireframe.shader]: Yeah without geometry shaders the wireframe view is not possible unless you do computations on the CPU. 

Is there a way to disable the view if the GPU doesn't suport 4.1?

Comment (fieldOfView) [inline plugins/WireframeView/wireframe.shader]: See OpenGLContext.detectBestOpenGLVersion()
I think the best way would be to use that to check for the right OpenGL version in the __init__.py register() method and return an empty dict when the gl version is too low.

**Comment by awhiemstra** [comment]
There is actually a very simple trick to do the same on all version of GL (and not just 4.1+): https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glPolygonMode.xhtml

Also, we used to have a wireframe view back in the day, but at some point we removed it since it did not have a lot of purpose and it did not work anymore. I still see little reason to have it personally.

**Comment by BagelOrb** [comment]
I was just trying some stuff out. If you think it shouldn't be included
than we can just leave it out. Maybe it's a nice one for the online plugin
repository. Nobody likes a sparsely populated plugin repository ;)

Op 22 mei 2017 16:59 schreef "awhiemstra" <notifications@github.com>:

**Comment by LipuFei** [comment]
Expect for the OpenGL discussion, the rest looks fine. I have also added a legend panel for the WireframeView. **I have checked that the colors are correct, but they somehow look different on the legend and the build plate. @Ghostkeeper and I think it could be the lighting.**

image

File: plugins/WireframeView/WireframeView.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        self._wireframe_shader = None
+
+    def beginRendering(self):
+        scene = self.getController().getScene()
+        renderer = self.getRenderer()
+
+        if not self._wireframe_shader:
+            self._wireframe_shader = OpenGL.getInstance().createShaderProgram(os.path.join(PluginRegistry.getInstance().getPluginPath("WireframeView"), "wireframe.shader"))
+            self._wireframe_shader.setUniformValue("u_color", Color(*Application.getInstance().getTheme().getColor("xray").getRgbF()))
```
Comment (fieldOfView) [inline plugins/WireframeView/WireframeView.py]: This is not very semantically correct, is it? This has nothing to do with "xray".

**Comment by nallath** [comment]
I think we should keep the plugin seperate; It doesn't add any interesting or particularly usefull features. This is a perfect candidate once we have the plugin repo / online thing up & running.

File: plugins/WireframeView/WireframeView.py (hunk 2 above)
Comment (LipuFei) [inline plugins/WireframeView/WireframeView.py]: I will create a separate set of colours for the Wireframe :)

**Comment by LipuFei** [comment]
@fieldOfView @BagelOrb says he doesn't know what this view is for :P

**Pull request closed without merging**